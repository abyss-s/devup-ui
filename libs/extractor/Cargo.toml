[package]
name = "extractor"
version = "0.1.0"
edition = "2024"

[dependencies]
oxc_parser = "0.83.0"
oxc_syntax = "0.83.0"
oxc_span = "0.83.0"
oxc_allocator = "0.83.0"
oxc_ast = "0.83.0"
oxc_ast_visit = "0.83.0"
oxc_codegen = "0.83.0"
css = { path = "../css" }
phf = "0.13"
strum = "0.27.2"
strum_macros = "0.27.2"

[dev-dependencies]
insta = "1.43.1"
serial_test = "3.2.0"
rstest = "0.26.1"
