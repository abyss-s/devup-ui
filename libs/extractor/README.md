## Extractor

jsx to css extractor

### Example

Before

```jsx
<Box bg="red" color="white" p={2} m={2} textAlign={variable}>
    Hello World
</Box>
```

After

```jsx
<Box className={"dp1 dp2 dp-3 dp-4 dp-v"} style={{
    "--ta": variable,
}}>
    Hello World
</Box>
```

```mermaid
graph TD
;
    Code --> ExtractStyleProp;
    ExtractStyleProp --> ExtractStyleValue;
    ExtractStyleValue --> GenCssFile;
    ExtractStyleValue --> EditCode;
    ExtractStyleProp --> EditCode;
```