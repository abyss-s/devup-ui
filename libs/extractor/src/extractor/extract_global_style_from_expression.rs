use std::collections::BTreeMap;

use crate::{
    ExtractStyleProp,
    css_utils::css_to_style,
    extract_style::{
        extract_font_face::ExtractFontFace, extract_import::ExtractImport,
        extract_style_value::ExtractStyleValue,
    },
    extractor::{
        GlobalExtractResult, extract_style_from_expression::extract_style_from_expression,
    },
    utils::get_string_by_literal_expression,
};
use css::{
    disassemble_property,
    optimize_multi_css_value::{check_multi_css_optimize, optimize_mutli_css_value},
    style_selector::StyleSelector,
};
use oxc_ast::{
    AstBuilder,
    ast::{ArrayExpressionElement, Expression, ObjectPropertyKind, PropertyKey},
};

pub fn extract_global_style_from_expression<'a>(
    ast_builder: &AstBuilder<'a>,
    expression: &mut Expression<'a>,
    file: &str,
) -> GlobalExtractResult<'a> {
    let mut styles = vec![];

    if let Expression::ObjectExpression(obj) = expression {
        for p in obj.properties.iter_mut() {
            if let ObjectPropertyKind::ObjectProperty(o) = p
                && let Some(name) = if let PropertyKey::StaticIdentifier(ident) = &o.key {
                    Some(ident.name.to_string())
                } else if let PropertyKey::StringLiteral(s) = &o.key {
                    Some(s.value.to_string())
                } else if let PropertyKey::TemplateLiteral(t) = &o.key {
                    Some(
                        t.quasis
                            .iter()
                            .map(|q| q.value.raw.as_str())
                            .collect::<String>(),
                    )
                } else {
                    None
                }
            {
                if name == "imports" {
                    if let Expression::ArrayExpression(arr) = &o.value {
                        for p in arr.elements.iter() {
                            if let Some(url) = if let ArrayExpressionElement::StringLiteral(s) = p {
                                Some(s.value.trim().to_string())
                            } else if let ArrayExpressionElement::TemplateLiteral(t) = p {
                                Some(
                                    t.quasis
                                        .iter()
                                        .map(|q| q.value.raw.as_str())
                                        .collect::<String>()
                                        .trim()
                                        .to_string(),
                                )
                            } else {
                                None
                            } {
                                styles.push(ExtractStyleProp::Static(ExtractStyleValue::Import(
                                    ExtractImport {
                                        url,
                                        file: file.to_string(),
                                    },
                                )));
                            }
                        }
                    }
                } else if name == "fontFaces" {
                    if let Expression::ArrayExpression(arr) = &o.value {
                        for p in arr.elements.iter() {
                            if let ArrayExpressionElement::ObjectExpression(o) = p {
                                styles.push(ExtractStyleProp::Static(ExtractStyleValue::FontFace(ExtractFontFace {
                                    properties: BTreeMap::from_iter(
                                        o.properties
                                            .iter()
                                            .filter_map(|p| {
                                                if let ObjectPropertyKind::ObjectProperty(o) = p
                                                    && let PropertyKey::StaticIdentifier(ident) = &o.key
                                                    && let Some(s) = get_string_by_literal_expression(&o.value)
                                                {
                                                    Some(disassemble_property(&ident.name).iter().map(|p| (p.to_string(), if check_multi_css_optimize(p) { optimize_mutli_css_value(&s) } else { s.clone() })).collect::<Vec<_>>())
                                                } else {
                                                    None
                                                }
                                            })
                                            .flatten(),
                                    ),
                                    file: file.to_string(),
                                })));
                            } else if let ArrayExpressionElement::TemplateLiteral(t) = p {
                                let css_styles = css_to_style(
                                    t.quasis
                                        .iter()
                                        .map(|q| q.value.raw.as_str())
                                        .collect::<String>()
                                        .trim(),
                                    0,
                                    &None,
                                )
                                .into_iter()
                                .map(ExtractStyleValue::Static)
                                .collect::<Vec<_>>();
                                styles.push(ExtractStyleProp::Static(ExtractStyleValue::FontFace(
                                    ExtractFontFace {
                                        properties: BTreeMap::from_iter(
                                            css_styles.iter().filter_map(|p| {
                                                if let ExtractStyleValue::Static(st) = p {
                                                    Some((
                                                        st.property().to_string(),
                                                        st.value().to_string(),
                                                    ))
                                                } else {
                                                    None
                                                }
                                            }),
                                        ),
                                        file: file.to_string(),
                                    },
                                )));
                            }
                        }
                    }
                } else {
                    styles.extend(
                        extract_style_from_expression(
                            ast_builder,
                            None,
                            &mut o.value,
                            0,
                            &Some(StyleSelector::Global(name.clone(), file.to_string())),
                        )
                        .styles,
                    );
                }
            }
        }
    }
    GlobalExtractResult {
        styles,
        style_order: None,
    }
}
