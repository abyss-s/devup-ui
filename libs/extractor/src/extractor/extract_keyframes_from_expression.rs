use crate::{
    ExtractStyleProp,
    extract_style::{extract_keyframes::ExtractKeyframes, extract_style_value::ExtractStyleValue},
    extractor::{
        KeyframesExtractResult, extract_style_from_expression::extract_style_from_expression,
    },
};
use oxc_ast::{
    AstBuilder,
    ast::{Expression, ObjectPropertyKind, PropertyKey},
};

pub fn extract_keyframes_from_expression<'a>(
    ast_builder: &AstBuilder<'a>,
    expression: &mut Expression<'a>,
) -> KeyframesExtractResult {
    let mut keyframes = ExtractKeyframes::default();

    if let Expression::ObjectExpression(obj) = expression {
        for p in obj.properties.iter_mut() {
            if let ObjectPropertyKind::ObjectProperty(o) = p {
                let mut name = if let PropertyKey::StaticIdentifier(ident) = &o.key {
                    ident.name.to_string()
                } else if let PropertyKey::StringLiteral(s) = &o.key {
                    s.value.to_string()
                } else if let PropertyKey::TemplateLiteral(t) = &o.key {
                    t.quasis
                        .iter()
                        .map(|q| q.value.raw.as_str())
                        .collect::<Vec<_>>()
                        .join("")
                } else if let PropertyKey::NumericLiteral(n) = &o.key {
                    n.value.to_string()
                } else {
                    continue;
                };
                // convert number
                if let Ok(num) = name.parse::<f64>() {
                    name = format!("{num}%");
                }
                let mut styles =
                    extract_style_from_expression(ast_builder, None, &mut o.value, 0, None)
                        .styles
                        .into_iter()
                        .filter_map(|s| match s {
                            ExtractStyleProp::Static(ExtractStyleValue::Static(s)) => Some(s),
                            _ => None,
                        })
                        .collect::<Vec<_>>();
                styles.sort_by_key(|a| a.property().to_string());
                keyframes.keyframes.insert(name, styles);
            }
        }
    }
    KeyframesExtractResult { keyframes }
}
