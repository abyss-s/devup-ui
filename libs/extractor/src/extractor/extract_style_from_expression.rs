use crate::{
    ExtractStyleProp,
    css_utils::css_to_style,
    extract_style::{
        extract_dynamic_style::ExtractDynamicStyle, extract_static_style::ExtractStaticStyle,
        extract_style_value::ExtractStyleValue,
    },
    extractor::{
        ExtractResult, extract_style_from_member_expression::extract_style_from_member_expression,
    },
    utils::{
        expression_to_code, get_number_by_literal_expression, get_string_by_literal_expression,
        is_same_expression, is_special_property,
    },
};
use css::style_selector::StyleSelector;
use oxc_allocator::CloneIn;
use oxc_ast::{
    AstBuilder,
    ast::{
        BinaryOperator, Expression, LogicalOperator, ObjectPropertyKind, PropertyKey,
        TemplateElementValue,
    },
};
use oxc_span::SPAN;

const IGNORED_IDENTIFIERS: [&str; 3] = ["undefined", "NaN", "Infinity"];

pub fn extract_style_from_expression<'a>(
    ast_builder: &AstBuilder<'a>,
    name: Option<&str>,
    expression: &mut Expression<'a>,
    level: u8,
    selector: Option<&StyleSelector>,
) -> ExtractResult<'a> {
    let mut typo = false;

    if name.is_none() && selector.is_none() {
        let mut style_order = None;
        let mut style_vars = None;
        return match expression {
            Expression::ObjectExpression(obj) => {
                let mut props_styles: Vec<ExtractStyleProp<'_>> = vec![];
                let mut tag = None;
                for idx in (0..obj.properties.len()).rev() {
                    let mut prop = obj.properties.remove(idx);
                    if !match &mut prop {
                        ObjectPropertyKind::ObjectProperty(prop) => {
                            if let PropertyKey::StaticIdentifier(ident) = &prop.key
                                && let name = ident.name.as_str()
                                && !is_special_property(name)
                            {
                                if name == "styleOrder" {
                                    style_order = get_number_by_literal_expression(&prop.value)
                                        .map(|v| v as u8);
                                    continue;
                                }
                                if name == "styleVars" {
                                    style_vars = Some(prop.value.clone_in(ast_builder.allocator));
                                    continue;
                                }

                                let ExtractResult {
                                    styles, tag: _tag, ..
                                } = extract_style_from_expression(
                                    ast_builder,
                                    Some(name),
                                    &mut prop.value,
                                    0,
                                    None,
                                );
                                props_styles.extend(styles);
                                tag = _tag.or(tag);
                                true
                            } else {
                                false
                            }
                        }
                        ObjectPropertyKind::SpreadProperty(prop) => {
                            let ExtractResult {
                                styles, tag: _tag, ..
                            } = extract_style_from_expression(
                                ast_builder,
                                None,
                                &mut prop.argument,
                                0,
                                None,
                            );
                            props_styles.extend(styles);
                            tag = _tag.or(tag);
                            false
                        }
                    } {
                        obj.properties.insert(idx, prop);
                    }
                }
                ExtractResult {
                    styles: props_styles,
                    tag,
                    style_order,
                    style_vars,
                }
            }
            Expression::ConditionalExpression(conditional) => ExtractResult {
                styles: vec![ExtractStyleProp::Conditional {
                    condition: conditional.test.clone_in(ast_builder.allocator),
                    consequent: Some(Box::new(ExtractStyleProp::StaticArray(
                        extract_style_from_expression(
                            ast_builder,
                            None,
                            &mut conditional.consequent,
                            level,
                            None,
                        )
                        .styles,
                    ))),
                    alternate: Some(Box::new(ExtractStyleProp::StaticArray(
                        extract_style_from_expression(
                            ast_builder,
                            None,
                            &mut conditional.alternate,
                            level,
                            selector,
                        )
                        .styles,
                    ))),
                }],
                tag: None,
                style_order,
                style_vars,
            },
            Expression::ParenthesizedExpression(parenthesized) => extract_style_from_expression(
                ast_builder,
                None,
                &mut parenthesized.expression,
                level,
                None,
            ),
            _ => ExtractResult::default(),
        };
    }

    if let Some(name) = name {
        if is_special_property(name) {
            return ExtractResult::default();
        }

        if name == "as" {
            return ExtractResult {
                tag: Some(expression.clone_in(ast_builder.allocator)),
                ..ExtractResult::default()
            };
        }
        if name == "selectors"
            && let Expression::ObjectExpression(obj) = expression
        {
            let mut props = vec![];
            for p in obj.properties.iter_mut() {
                if let ObjectPropertyKind::ObjectProperty(o) = p {
                    let name = o.key.name().unwrap().to_string();
                    props.extend(
                        extract_style_from_expression(
                            ast_builder,
                            None,
                            &mut o.value,
                            level,
                            Some(
                                &if let Some(selector) = selector {
                                    name.replace("&", &selector.to_string())
                                } else {
                                    name
                                }
                                .as_str()
                                .into(),
                            ),
                        )
                        .styles,
                    );
                }
            }
            return ExtractResult {
                styles: props,
                ..ExtractResult::default()
            };
        }

        if let Some(new_selector) = name.strip_prefix("_") {
            return extract_style_from_expression(
                ast_builder,
                None,
                expression,
                level,
                Some(&if let Some(selector) = selector {
                    (selector, new_selector).into()
                } else {
                    new_selector.into()
                }),
            );
        }
        typo = name == "typography";
    }
    if let Some(value) = get_string_by_literal_expression(expression) {
        if let Some(name) = name {
            ExtractResult {
                styles: vec![ExtractStyleProp::Static(if typo {
                    ExtractStyleValue::Typography(value.to_string())
                } else {
                    ExtractStyleValue::Static(ExtractStaticStyle::new(
                        name,
                        &value,
                        level,
                        selector.cloned(),
                    ))
                })],
                ..ExtractResult::default()
            }
        } else {
            ExtractResult {
                styles: css_to_style(&value, level, &selector),
                ..ExtractResult::default()
            }
        }
    } else {
        match expression {
            Expression::UnaryExpression(_)
            | Expression::BinaryExpression(_)
            | Expression::StaticMemberExpression(_)
            | Expression::CallExpression(_) => ExtractResult {
                styles: vec![ExtractStyleProp::Static(ExtractStyleValue::Dynamic(
                    ExtractDynamicStyle::new(
                        name.unwrap(),
                        level,
                        &expression_to_code(expression),
                        selector.cloned(),
                    ),
                ))],
                ..ExtractResult::default()
            },
            Expression::TSAsExpression(exp) => extract_style_from_expression(
                ast_builder,
                name,
                &mut exp.expression,
                level,
                selector,
            ),
            Expression::ComputedMemberExpression(mem) => {
                extract_style_from_member_expression(ast_builder, name, mem, level, selector)
            }
            Expression::TemplateLiteral(tmp) => {
                if let Some(name) = name {
                    if tmp.quasis.len() == 1 {
                        ExtractResult {
                            styles: vec![ExtractStyleProp::Static(if typo {
                                ExtractStyleValue::Typography(tmp.quasis[0].value.raw.to_string())
                            } else {
                                ExtractStyleValue::Static(ExtractStaticStyle::new(
                                    name,
                                    &tmp.quasis[0].value.raw,
                                    level,
                                    selector.cloned(),
                                ))
                            })],
                            ..ExtractResult::default()
                        }
                    } else if typo {
                        ExtractResult {
                            styles: vec![ExtractStyleProp::Expression {
                                expression: ast_builder.expression_template_literal(
                                    SPAN,
                                    ast_builder.vec_from_array([
                                        ast_builder.template_element(
                                            SPAN,
                                            TemplateElementValue {
                                                raw: ast_builder.atom("typo-"),
                                                cooked: None,
                                            },
                                            false,
                                        ),
                                        ast_builder.template_element(
                                            SPAN,
                                            TemplateElementValue {
                                                raw: ast_builder.atom(""),
                                                cooked: None,
                                            },
                                            true,
                                        ),
                                    ]),
                                    ast_builder.vec_from_array([
                                        expression.clone_in(ast_builder.allocator)
                                    ]),
                                ),
                                styles: vec![],
                            }],
                            ..ExtractResult::default()
                        }
                    } else {
                        ExtractResult {
                            styles: vec![ExtractStyleProp::Static(ExtractStyleValue::Dynamic(
                                ExtractDynamicStyle::new(
                                    name,
                                    level,
                                    &expression_to_code(expression),
                                    selector.cloned(),
                                ),
                            ))],
                            ..ExtractResult::default()
                        }
                    }
                } else {
                    ExtractResult::default()
                }
            }
            Expression::Identifier(identifier) => {
                if IGNORED_IDENTIFIERS.contains(&identifier.name.as_str()) {
                    ExtractResult::default()
                } else if let Some(name) = name {
                    if typo {
                        ExtractResult {
                            styles: vec![ExtractStyleProp::Expression {
                                expression: ast_builder.expression_conditional(
                                    SPAN,
                                    ast_builder
                                        .expression_identifier(SPAN, identifier.name.as_str()),
                                    ast_builder.expression_template_literal(
                                        SPAN,
                                        ast_builder.vec_from_array([
                                            ast_builder.template_element(
                                                SPAN,
                                                TemplateElementValue {
                                                    raw: ast_builder.atom("typo-"),
                                                    cooked: None,
                                                },
                                                false,
                                            ),
                                            ast_builder.template_element(
                                                SPAN,
                                                TemplateElementValue {
                                                    raw: ast_builder.atom(""),
                                                    cooked: None,
                                                },
                                                true,
                                            ),
                                        ]),
                                        ast_builder.vec_from_array([
                                            expression.clone_in(ast_builder.allocator)
                                        ]),
                                    ),
                                    ast_builder.expression_string_literal(SPAN, "", None),
                                ),
                                styles: vec![],
                            }],
                            ..ExtractResult::default()
                        }
                    } else {
                        ExtractResult {
                            styles: vec![ExtractStyleProp::Static(ExtractStyleValue::Dynamic(
                                ExtractDynamicStyle::new(
                                    name,
                                    level,
                                    &identifier.name,
                                    selector.cloned(),
                                ),
                            ))],
                            ..ExtractResult::default()
                        }
                    }
                } else {
                    ExtractResult::default()
                }
            }
            Expression::LogicalExpression(logical) => {
                let res = Some(Box::new(ExtractStyleProp::StaticArray(
                    extract_style_from_expression(
                        ast_builder,
                        name,
                        &mut logical.right,
                        level,
                        selector,
                    )
                    .styles,
                )));
                match logical.operator {
                    LogicalOperator::Or => ExtractResult {
                        styles: vec![ExtractStyleProp::Conditional {
                            condition: logical.left.clone_in(ast_builder.allocator),
                            consequent: None,
                            alternate: res,
                        }],
                        ..ExtractResult::default()
                    },
                    LogicalOperator::And => ExtractResult {
                        styles: vec![ExtractStyleProp::Conditional {
                            condition: logical.left.clone_in(ast_builder.allocator),
                            consequent: res,
                            alternate: None,
                        }],
                        ..ExtractResult::default()
                    },
                    LogicalOperator::Coalesce => ExtractResult {
                        styles: vec![ExtractStyleProp::Conditional {
                            condition: ast_builder.expression_logical(
                                SPAN,
                                ast_builder.expression_binary(
                                    SPAN,
                                    logical.left.clone_in(ast_builder.allocator),
                                    BinaryOperator::StrictInequality,
                                    ast_builder.expression_null_literal(SPAN),
                                ),
                                LogicalOperator::And,
                                ast_builder.expression_binary(
                                    SPAN,
                                    logical.left.clone_in(ast_builder.allocator),
                                    BinaryOperator::StrictInequality,
                                    ast_builder.expression_identifier(SPAN, "undefined"),
                                ),
                            ),
                            consequent: Some(Box::new(ExtractStyleProp::StaticArray(
                                extract_style_from_expression(
                                    ast_builder,
                                    name,
                                    &mut logical.left,
                                    level,
                                    selector,
                                )
                                .styles,
                            ))),
                            alternate: res,
                        }],
                        ..ExtractResult::default()
                    },
                }
            }
            Expression::ParenthesizedExpression(parenthesized) => extract_style_from_expression(
                ast_builder,
                name,
                &mut parenthesized.expression,
                level,
                selector,
            ),
            Expression::ArrayExpression(array) => {
                let mut props = vec![];

                for (idx, element) in array.elements.iter_mut().enumerate() {
                    props.extend(
                        extract_style_from_expression(
                            ast_builder,
                            name,
                            element.to_expression_mut(),
                            idx as u8,
                            selector,
                        )
                        .styles,
                    );
                }
                ExtractResult {
                    styles: vec![ExtractStyleProp::StaticArray(props)],
                    tag: None,
                    style_order: None,
                    style_vars: None,
                }
            }
            Expression::ConditionalExpression(conditional) => {
                if is_same_expression(&conditional.consequent, &conditional.alternate) {
                    extract_style_from_expression(
                        ast_builder,
                        name,
                        &mut conditional.consequent,
                        level,
                        selector,
                    )
                } else {
                    ExtractResult {
                        styles: vec![ExtractStyleProp::Conditional {
                            condition: conditional.test.clone_in(ast_builder.allocator),
                            consequent: Some(Box::new(ExtractStyleProp::StaticArray(
                                extract_style_from_expression(
                                    ast_builder,
                                    name,
                                    &mut conditional.consequent,
                                    level,
                                    selector,
                                )
                                .styles,
                            ))),
                            alternate: Some(Box::new(ExtractStyleProp::StaticArray(
                                extract_style_from_expression(
                                    ast_builder,
                                    name,
                                    &mut conditional.alternate,
                                    level,
                                    selector,
                                )
                                .styles,
                            ))),
                        }],
                        ..ExtractResult::default()
                    }
                }
            }
            Expression::ObjectExpression(obj) => {
                let mut props = vec![];
                for p in obj.properties.iter_mut() {
                    if let ObjectPropertyKind::ObjectProperty(o) = p {
                        props.extend(
                            extract_style_from_expression(
                                ast_builder,
                                Some(&o.key.name().unwrap()),
                                &mut o.value,
                                level,
                                selector,
                            )
                            .styles,
                        );
                    }
                }
                ExtractResult {
                    styles: props,
                    ..ExtractResult::default()
                }
            }
            _ => ExtractResult::default(),
        }
    }
}
