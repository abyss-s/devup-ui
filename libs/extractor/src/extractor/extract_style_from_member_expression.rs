use crate::{
    ExtractStyleProp,
    extract_style::{
        extract_dynamic_style::ExtractDynamicStyle, extract_style_value::ExtractStyleValue,
    },
    extractor::{ExtractResult, extract_style_from_expression::extract_style_from_expression},
    utils::{
        expression_to_code, get_number_by_literal_expression, get_string_by_literal_expression,
    },
};
use css::style_selector::StyleSelector;
use oxc_allocator::CloneIn;
use oxc_ast::{
    AstBuilder,
    ast::{
        ArrayExpressionElement, ComputedMemberExpression, Expression, ObjectPropertyKind,
        PropertyKey,
    },
};
use oxc_span::SPAN;
use std::collections::BTreeMap;

pub(super) fn extract_style_from_member_expression<'a>(
    ast_builder: &AstBuilder<'a>,
    name: Option<&str>,
    mem: &mut ComputedMemberExpression<'a>,
    level: u8,
    selector: &Option<StyleSelector>,
) -> ExtractResult<'a> {
    let mem_expression = &mem.expression.clone_in(ast_builder.allocator);
    let mut ret: Vec<ExtractStyleProp> = vec![];

    match &mut mem.object {
        Expression::ArrayExpression(array) => {
            if array.elements.is_empty() {
                return ExtractResult::default();
            }

            if let Some(num) = get_number_by_literal_expression(mem_expression) {
                if num < 0f64 {
                    return ExtractResult::default();
                }
                let mut etc = None;
                for (idx, p) in array.elements.iter_mut().enumerate() {
                    if let ArrayExpressionElement::SpreadElement(sp) = p {
                        etc = Some(sp.argument.clone_in(ast_builder.allocator));
                        continue;
                    }
                    if idx as f64 == num {
                        return extract_style_from_expression(
                            ast_builder,
                            name,
                            p.to_expression_mut(),
                            level,
                            selector,
                        );
                    }
                }
                return ExtractResult {
                    styles: etc
                        .map(|etc| {
                            vec![ExtractStyleProp::Static(ExtractStyleValue::Dynamic(
                                ExtractDynamicStyle::new(
                                    name.unwrap(),
                                    level,
                                    &expression_to_code(&Expression::ComputedMemberExpression(
                                        ast_builder.alloc_computed_member_expression(
                                            SPAN,
                                            etc,
                                            mem_expression.clone_in(ast_builder.allocator),
                                            false,
                                        ),
                                    )),
                                    selector.clone(),
                                ),
                            ))]
                        })
                        .unwrap_or_default(),
                    tag: None,
                    style_order: None,
                    style_vars: None,
                };
            }

            let mut map = BTreeMap::new();
            for (idx, p) in array.elements.iter_mut().enumerate() {
                if let ArrayExpressionElement::SpreadElement(sp) = p {
                    map.insert(
                        idx.to_string(),
                        Box::new(ExtractStyleProp::Static(ExtractStyleValue::Dynamic(
                            ExtractDynamicStyle::new(
                                name.unwrap(),
                                level,
                                &expression_to_code(&Expression::ComputedMemberExpression(
                                    ast_builder.alloc_computed_member_expression(
                                        SPAN,
                                        sp.argument.clone_in(ast_builder.allocator),
                                        mem_expression.clone_in(ast_builder.allocator),
                                        false,
                                    ),
                                )),
                                selector.clone(),
                            ),
                        ))),
                    );
                } else {
                    map.insert(
                        idx.to_string(),
                        Box::new(ExtractStyleProp::StaticArray(
                            extract_style_from_expression(
                                ast_builder,
                                name,
                                p.to_expression_mut(),
                                level,
                                selector,
                            )
                            .styles,
                        )),
                    );
                }
            }

            ret.push(ExtractStyleProp::MemberExpression {
                expression: mem_expression.clone_in(ast_builder.allocator),
                map,
            });
        }
        Expression::ObjectExpression(obj) => {
            if obj.properties.is_empty() {
                return ExtractResult::default();
            }

            let mut map = BTreeMap::new();
            if let Some(k) = get_string_by_literal_expression(mem_expression) {
                let mut etc = None;
                for p in obj.properties.iter_mut() {
                    if let ObjectPropertyKind::ObjectProperty(o) = p {
                        if let PropertyKey::StaticIdentifier(ref pk) = o.key
                            && pk.name == k
                        {
                            return ExtractResult {
                                styles: extract_style_from_expression(
                                    ast_builder,
                                    name,
                                    &mut o.value,
                                    level,
                                    selector,
                                )
                                .styles,
                                ..ExtractResult::default()
                            };
                        }
                    } else if let ObjectPropertyKind::SpreadProperty(sp) = p {
                        etc = Some(sp.argument.clone_in(ast_builder.allocator));
                    }
                }

                match etc {
                    None => {
                        return ExtractResult::default();
                    }
                    Some(etc) => ret.push(ExtractStyleProp::Static(ExtractStyleValue::Dynamic(
                        ExtractDynamicStyle::new(
                            name.unwrap(),
                            level,
                            &expression_to_code(&Expression::ComputedMemberExpression(
                                ast_builder.alloc_computed_member_expression(
                                    SPAN,
                                    etc,
                                    mem_expression.clone_in(ast_builder.allocator),
                                    false,
                                ),
                            )),
                            selector.clone(),
                        ),
                    ))),
                }
            }

            for p in obj.properties.iter_mut() {
                if let ObjectPropertyKind::ObjectProperty(o) = p
                    && let PropertyKey::StaticIdentifier(_)
                    | PropertyKey::NumericLiteral(_)
                    | PropertyKey::StringLiteral(_) = o.key
                {
                    map.insert(
                        o.key.name().unwrap().to_string(),
                        Box::new(ExtractStyleProp::StaticArray(
                            extract_style_from_expression(
                                ast_builder,
                                name,
                                &mut o.value,
                                level,
                                selector,
                            )
                            .styles,
                        )),
                    );
                }
            }
            ret.push(ExtractStyleProp::MemberExpression {
                expression: mem_expression.clone_in(ast_builder.allocator),
                map,
            });
        }
        Expression::Identifier(_) => {
            if let Some(name) = name {
                ret.push(ExtractStyleProp::Static(ExtractStyleValue::Dynamic(
                    ExtractDynamicStyle::new(
                        name,
                        level,
                        &expression_to_code(&Expression::ComputedMemberExpression(
                            ast_builder.alloc_computed_member_expression(
                                SPAN,
                                mem.object.clone_in(ast_builder.allocator),
                                mem_expression.clone_in(ast_builder.allocator),
                                false,
                            ),
                        )),
                        selector.clone(),
                    ),
                )))
            }
        }
        _ => {}
    };

    ExtractResult {
        styles: ret,
        ..ExtractResult::default()
    }
}
