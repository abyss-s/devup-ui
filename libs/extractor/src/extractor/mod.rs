use oxc_ast::ast::Expression;

use crate::{ExtractStyleProp, extract_style::extract_keyframes::ExtractKeyframes};

pub(super) mod extract_global_style_from_expression;
pub(super) mod extract_keyframes_from_expression;
pub(super) mod extract_style_from_expression;
pub(super) mod extract_style_from_jsx;
pub(super) mod extract_style_from_member_expression;

/**
 * type
 * 1. jsx -> <Extract a={1} />
 * 2. object -> createElement('div', {a: 1})
 * 3. object with select -> createElement('div', {a: 1})
 */

#[derive(Debug, Default)]
pub struct ExtractResult<'a> {
    // attribute will be maintained
    pub styles: Vec<ExtractStyleProp<'a>>,
    pub tag: Option<Expression<'a>>,
    pub style_order: Option<u8>,
    pub style_vars: Option<Expression<'a>>,
}

#[derive(Debug)]
pub struct GlobalExtractResult<'a> {
    pub styles: Vec<ExtractStyleProp<'a>>,
    pub style_order: Option<u8>,
}

#[derive(Debug)]
pub struct KeyframesExtractResult {
    pub keyframes: ExtractKeyframes,
}
