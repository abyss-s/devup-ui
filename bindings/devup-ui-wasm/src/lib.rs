use extractor::{extract, ExtractOption, ExtractStyleValue, StyleProperty};
use once_cell::sync::Lazy;
use sheet::StyleSheet;
use std::sync::Mutex;
use wasm_bindgen::prelude::*;

static GLOBAL_STYLE_SHEET: Lazy<Mutex<StyleSheet>> = Lazy::new(|| Mutex::new(StyleSheet::new()));

#[wasm_bindgen]
pub struct Output {
    code: String,
    styles: Vec<ExtractStyleValue>,
}
#[wasm_bindgen]
extern "C" {
    #[wasm_bindgen(js_namespace = console)]
    fn log(s: &str);
}

#[wasm_bindgen]
impl Output {
    /// Get the code
    #[wasm_bindgen(getter)]
    pub fn code(&self) -> String {
        self.code.clone()
    }

    /// Get the css
    #[wasm_bindgen(getter)]
    pub fn css(&self) -> Option<String> {
        let mut sheet = GLOBAL_STYLE_SHEET.lock().unwrap();
        let mut collected = vec![];
        for style in self.styles.iter() {
            let (cls, variable) = match style.extract() {
                StyleProperty::ClassName(cls) => (cls, None),
                StyleProperty::Variable {
                    class_name,
                    variable_name,
                    ..
                } => (class_name, Some(variable_name)),
            };
            match style {
                ExtractStyleValue::Static(st) => {
                    if let Some(css) =
                        sheet.add_property(cls, st.property.clone(), st.value.clone())
                    {
                        collected.push(css);
                    }
                }
                ExtractStyleValue::Dynamic(dy) => {
                    if let Some(css) = sheet.add_property(
                        cls,
                        dy.property.clone(),
                        format!("var({})", variable.unwrap()),
                    ) {
                        collected.push(css);
                    }
                }
                ExtractStyleValue::Css(cs) => {
                    if let Some(css) = sheet.add_css(cls, cs.css.clone()) {
                        collected.push(css);
                    }
                }
            }
        }
        if collected.is_empty() {
            return None;
        }
        collected.push("".to_string());
        Some(collected.join("\n"))
    }
}

#[wasm_bindgen(js_name = "codeExtract", catch)]
pub fn code_extract(
    filename: &str,
    code: &str,
    package: &str,
    css_file: &str,
) -> Result<Output, JsValue> {
    match extract(
        filename,
        code,
        ExtractOption {
            package: package.to_string(),
            css_file: Some(css_file.to_string()),
        },
    ) {
        Ok(output) => Ok(Output {
            code: if output.styles.is_empty() {
                code.to_string()
            } else {
                output.code
            },
            styles: output.styles,
        }),
        Err(error) => Err(JsValue::from_str(error.to_string().as_str())),
    }
}
