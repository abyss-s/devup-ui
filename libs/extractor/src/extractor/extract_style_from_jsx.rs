use crate::extractor::{
    ExtractResult, extract_style_from_expression::extract_style_from_expression,
};
use oxc_allocator::CloneIn;
use oxc_ast::{
    AstBuilder,
    ast::{Expression, JSXAttributeValue},
};

pub fn extract_style_from_jsx<'a>(
    ast_builder: &AstBuilder<'a>,
    name: &str,
    value: &mut JSXAttributeValue<'a>,
) -> ExtractResult<'a> {
    match value {
        JSXAttributeValue::ExpressionContainer(expression) => expression
            .expression
            .as_expression()
            .map(|expression| expression.clone_in(ast_builder.allocator)),
        JSXAttributeValue::StringLiteral(literal) => Some(Expression::StringLiteral(
            literal.clone_in(ast_builder.allocator),
        )),
        _ => None,
    }
    .map(|mut expression| {
        extract_style_from_expression(ast_builder, Some(name), &mut expression, 0, &None)
    })
    .unwrap_or_default()
}
