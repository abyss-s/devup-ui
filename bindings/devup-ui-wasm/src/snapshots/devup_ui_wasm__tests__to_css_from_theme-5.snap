---
source: bindings/devup-ui-wasm/src/lib.rs
expression: theme.to_css()
---
":root{color-scheme:light;--primary:#000}:root[data-theme=a]{color-scheme:dark}:root[data-theme=b]{color-scheme:dark}:root[data-theme=c]{color-scheme:dark}"
